src/core/QCheck2.mli

```diff
@@ -0,0 +1,1992 @@
+      Example: [char_range 'a' 'z'] for all lower case ASCII letters.
+
+      Shrinks towards [origin] if specified, otherwise towards [low].
+
+      @raise Invalid_argument if [low > high].
+
+      @since 0.13 *)
+
+  (** {3 Choosing elements} *)
+
+  val oneof : 'a t list -> 'a t
+  (** [oneof l] constructs a generator that selects among the given list of generators [l].
+
+      Shrinks towards the first generator of the list.
+  *)
+
+  val oneofl : 'a list -> 'a t
+  (** [oneofl l] constructs a generator that selects among the given list of values [l].
+
+      Shrinks towards the first element of the list.
+  *)
+
+  val oneofa : 'a array -> 'a t
+  (** [oneofa a] constructs a generator that selects among the given array of values [a].
+
+      Shrinks towards the first element of the array.
+  *)
+
+  val frequency : (int * 'a t) list -> 'a t
+  (** Constructs a generator that selects among a given list of generators.
+      Each of the given generators are chosen based on a positive integer weight.
+
+      Shrinks towards the first element of the list.
+  *)
+
+  val frequencyl : (int * 'a) list -> 'a t
+  (** Constructs a generator that selects among a given list of values.
+      Each of the given values are chosen based on a positive integer weight.
+
+      Shrinks towards the first element of the list.
+  *)
+
+  val frequencya : (int * 'a) array -> 'a t
+  (** Constructs a generator that selects among a given array of values.
+      Each of the array entries are chosen based on a positive integer weight.
+
+      Shrinks towards the first element of the array.
+  *)
+
+  (** {3 Shuffling elements} *)
+
+  val shuffle_a : 'a array -> 'a array t
+  (** Returns a copy of the array with its elements shuffled. *)
+
+  val shuffle_l : 'a list -> 'a list t
+  (** Creates a generator of shuffled lists. *)
+
+  val shuffle_w_l : (int * 'a) list -> 'a list t
+  (** Creates a generator of weighted shuffled lists. A given list is shuffled on each
+      generation according to the weights of its elements. An element with a larger weight
+      is more likely to be at the front of the list than an element with a smaller weight.
+      If we want to pick random elements from the (head of) list but need to prioritize
+      some elements over others, this generator can be useful.
+
+      Example: given a weighted list [[1, "one"; 5, "five"; 10, "ten"]], the generator is
+      more likely to generate [["ten"; "five"; "one"]] or [["five"; "ten"; "one"]] than
+      [["one"; "ten"; "five"]] because "ten" and "five" have larger weights than "one".
+
+      @since 0.11
+  *)
+
+  (** {3 Corner cases} *)
+
+  val graft_corners : 'a t -> 'a list -> unit -> 'a t
+  (** [graft_corners gen l ()] makes a new generator that enumerates
+      the corner cases in [l] and then behaves like [g].
+
+      Does not shrink if the test fails on a grafted value.
+      Shrinks towards [gen] otherwise.
+
+      @since 0.6 *)
+
+  val int_pos_corners : int list
```

Comment:
maybe we could have `Corner` sub-module with simpler names inside.

Reply:
Are you strongly opinionated on this?
While I can do it, I feel like it doesn't belong here:
* it is unrelated to integrated shrinking
* this is another API breaking change

I don't want to be the one breaking every single wall :grin:

Reply:
it is related to integrated shrinking since it's a generator combinator :p. But no need to do it now indeed!